graphbolt/src/csc_sampling_graph.cc

```diff
@@ -256,8 +257,76 @@ inline torch::Tensor UniformPick(
       picked_neighbors =
           torch::randint(offset, offset + num_neighbors, {fanout}, options);
     } else {
-      picked_neighbors = torch::randperm(num_neighbors, options);
-      picked_neighbors = picked_neighbors.slice(0, 0, fanout) + offset;
+      picked_neighbors = torch::empty({fanout}, options);
+      AT_DISPATCH_INTEGRAL_TYPES(
+          picked_neighbors.scalar_type(), "UniformPick", ([&] {
+            scalar_t* picked_neighbors_data =
+                picked_neighbors.data_ptr<scalar_t>();
+            if (fanout < num_neighbors / 10) {
+              // If set of numbers is small (up to 64) use linear search to
+              // verify uniqueness this operation is cheaper for CPU.
+              if (fanout && fanout < 64) {
+                *picked_neighbors_data = RandomEngine::ThreadLocal()->RandInt(
+                    offset, offset + num_neighbors);
+                auto b = picked_neighbors_data + 1;
+                auto e = picked_neighbors_data + fanout;
+
+                while (b != e) {
+                  // Put the new random number in the last position.
+                  *b = RandomEngine::ThreadLocal()->RandInt(
+                      offset, offset + num_neighbors);
+                  // Check if a new value doesn't exist in current
+                  // range(picked_neighbors_data, b). Otherwise get a new
+                  // value until we haven't unique range of elements.
+                  auto it = std::find(picked_neighbors_data, b, *b);
+                  if (it != b) continue;
+                  ++b;
+                }
+
+              } else {
+                // Use hash set.
+                // In the best scenario, time complexity is O(fanout), i.e., no
+                // conflict.
+                //
+                // Let k be fanout / num_neighbors, the expected number of extra
+                // sampling steps is roughly k^2 / (1-k) * num_neighbors, which
+                // means in the worst case scenario, the time complexity is
+                // O(num_neighbors^2). In practice,
+                // we use 1/10 since std::unordered_set is pretty slow.
+                std::unordered_set<scalar_t> picked_set;
+                while (static_cast<int64_t>(picked_set.size()) < fanout) {
+                  picked_set.insert(RandomEngine::ThreadLocal()->RandInt(
+                      offset, offset + num_neighbors));
+                }
+                std::copy(
+                    picked_set.begin(), picked_set.end(),
+                    picked_neighbors_data);
+              }
+
+            } else {
+              // In this case, `fanout >= num_neighbors / 10`. To reduce the
+              // computation overhead, we should reduce the number of random
+              // number generations. Even though reservior algorithm is more
+              // memory effficient (it has O(fanout) memory complexity), it
```

Comment:
check spelling, try to use chatgpt to help you rephrase wording to make it shorter and easier to read.

Reply:
I migrated comments from `DGL`, so do I need to update the comments in the original source code?

Reply:
It's ok to leave the existing comment as it, but make sure our new code are high quality.